fix(core): viewport trigger deregistering callbacks multiple times

crisbeto opened the pull request:
Adds a check to the viewport cleanup function to prevent it from re-processing elements that have been fully cleaned up, because it can lead to the `IntersectionObserver` being destroyed even though there are still pending triggers. This can happen, because we have cleanup callbacks both for the block is loaded, but also when the placeholder view is destroyed.

Fixes #52113.

crisbeto commented on packages/core/src/render3/instructions/defer_events.ts:
```diff
@@ -176,6 +176,11 @@ class DeferIntersectionManager {
 
     return () => {
+      // It's possible that a different cleanup callback fully removed this element already.
+      if (!this.viewportTriggers.has(trigger)) {
```
Note: not quite sure how we can test this, because it's pretty deep in the runtime implementation.

AndrewKushnir replied:
We've recently created an [integration app](https://github.com/angular/angular/tree/main/integration/defer/e2e) where we can add e2e tests, would that help?

If we skip adding a test, can we add a bit more info (as a comment in the code) about a situation that may trigger this? We can also add a TODO for now and come back to this after the feature freeze as a part of the overall stabilization effort.

JoostK commented on packages/core/src/render3/instructions/defer_events.ts:
```diff
@@ -176,6 +176,11 @@ class DeferIntersectionManager {
 
     return () => {
+      // It's possible that a different cleanup callback fully removed this element already.
+      if (!this.viewportTriggers.has(trigger)) {
```
You could save some bytes using

```ts
if (!entry!.callbacks.delete(callback)) {
  return;
}
```

AndrewKushnir replied:
(I'd like to ask for a quick comment above the `if` in this case :))

crisbeto committed:
fix(core): viewport trigger deregistering callbacks multiple times

Adds a check to the viewport cleanup function to prevent it from re-processing elements that have been fully cleaned up, because it can lead to the `IntersectionObserver` being destroyed even though there are still pending triggers. This can happen, because we have cleanup callbacks both for the block is loaded, but also when the placeholder view is destroyed.

Fixes #52113.

crisbeto commented:
I found a way to capture it in a test.

crisbeto commented:
Caretaker note: the CI is passing, but it's not green because the CI is referring to a check that doesn't exist.
